File: 2021-08-13 (day26)/Husky/solution_1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        cycle = set()
        while head:
            if head in cycle:
                return True
            else:
                cycle.add(head)

            head = head.next
        return False

File: DataStructures/LinkedList/alpha/middle-of-the-linked-list/Solution1.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function (head) {
  //first find this list's length
  let node = head;

  let length = 0;

  while (node !== null) {
    length++;

    node = node.next;
  }

  if (length === 0) return null;

  //find middle and reset node is head to find current list
  let middle = Math.floor(length / 2);

  node = head;

  while (middle > 0) {
    node = node.next;

    middle--;
  }
  return node;
};
